### PR TITLE
Improve endpoint handling.

### PR DESCRIPTION
**Pull request by ObiWahn**
- [ ] add tests

### CODE REVIEW

File: arangosh/Shell/V8ClientConnection.cpp [hunk 1]
```diff
@@ -506,7 +506,7 @@ static void ClientConnection_reconnect(v8::FunctionCallbackInfo<v8::Value> const
   }
 
   V8SecurityFeature& v8security = v8connection->server().getFeature<V8SecurityFeature>();
-  if (!v8security.isAllowedToConnectToEndpoint(isolate, endpoint)) {
+  if (!v8security.isAllowedToConnectToEndpoint(isolate, endpoint, endpoint)) {
     TRI_V8_THROW_EXCEPTION_MESSAGE(TRI_ERROR_FORBIDDEN,
                                    "not allowed to connect to this endpoint");
```
Comment (Simran-B) [inline arangosh/Shell/V8ClientConnection.cpp]: Adding my change from #11232 here (will remove it from my PR):
```suggestion
                                   std::string("not allowed to connect to this endpoint")
                                   endpoint);
```

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8888/

**Comment by Simran-B** [comment]
There is an inconsistency between `ssl://` and `https://`, namely the port has to be explicitly added to the request in the latter case:

```js
arangosh --javascript.endpoints-whitelist "ssl://www\.arangodb\.com:443\$"
require("internal").download("https://www.arangodb.com")     // allowed
require("internal").download("https://www.arangodb.com:443") // allowed

arangosh --javascript.endpoints-whitelist "https://www\.arangodb\.com:443\$"
require("internal").download("https://www.arangodb.com")     // NOT allowed!
require("internal").download("https://www.arangodb.com:443") // allowed
```

Not directly related, but odd: there is a pseudo-protocol(?) `src://` for which we seem to omit the protocol and port internally?!

```json
127.0.0.1:8529@_system> require("internal").download("http://asdf.foo/path")
{
  "code" : 500,
  "message" : "Could not connect to 'http+tcp://asdf.foo:80' 'getaddrinfo for host 'asdf.foo': Name does not resolve'"
}
127.0.0.1:8529@_system> require("internal").download("srv://asdf.foo/path")
{
  "code" : 500,
  "message" : "Could not connect to 'asdf.foo' ''"
}
```

**Comment by ObiWahn** [comment]
> There is an inconsistency between `ssl://` and `https://`, namely the port has to be explicitly added to the request in the latter case:
> 
> ```js
> arangosh --javascript.endpoints-whitelist "ssl://www\.arangodb\.com:443\$"
> require("internal").download("https://www.arangodb.com")     // allowed
> require("internal").download("https://www.arangodb.com:443") // allowed
> 
> arangosh --javascript.endpoints-whitelist "https://www\.arangodb\.com:443\$"
> require("internal").download("https://www.arangodb.com")     // NOT allowed!
> require("internal").download("https://www.arangodb.com:443") // allowed
> ```
> 

You could make the port optional in the 2nd regular expression! The first rule apparently works with the endpoint check only and is important to be backward compatible. We should really encourage users just to use the `http/https` variants and deprecate other combinations like `udp+vst`.

> Not directly related, but odd: there is a pseudo-protocol(?) `src://` for which we seem to omit the protocol and port internally?!
> 
> ```json
> 127.0.0.1:8529@_system> require("internal").download("http://asdf.foo/path")
> {
>   "code" : 500,
>   "message" : "Could not connect to 'http+tcp://asdf.foo:80' 'getaddrinfo for host 'asdf.foo': Name does not resolve'"
> }
> 127.0.0.1:8529@_system> require("internal").download("srv://asdf.foo/path")
> {
>   "code" : 500,
>   "message" : "Could not connect to 'asdf.foo' ''"
> }
> ```

The failure shown here is not an permission error. Therefor it will not be fixed in this PR.

File: lib/V8/v8-utils.cpp [hunk 2]
```diff
@@ -659,6 +659,105 @@ static std::string GetEndpointFromUrl(std::string const& url) {
...
+         relative = url.substr(6 + endpoint.length());
+
+         if (relative.empty() || relative[0] != '/') {
+           relative = "/" + relative;
+         }
+         if (endpoint.find(':') == std::string::npos) {
+           endpoint.append(":80");
+         }
+         endpoint = "tcp://" + endpoint;
+       } else if (url.substr(0, 6) == "srv://") {
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]: Indentation of whole block is wrong.

File: lib/V8/v8-utils.cpp [hunk 3]
```diff
@@ -659,6 +659,105 @@ static std::string GetEndpointFromUrl(std::string const& url) {
...
+         endpoint = GetEndpointFromUrl(url).substr(6);
+         relative = url.substr(6 + endpoint.length());
+
+         if (relative.empty() || relative[0] != '/') {
+           relative = "/" + relative;
+         }
+         if (endpoint.find(':') == std::string::npos) {
+           endpoint.append(":80");
+         }
+         endpoint = "tcp://" + endpoint;
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]: shouldn't this be `endpoint = "ssl://" + endpoint;`?

File: lib/V8/v8-utils.cpp [hunk 4]
```diff
@@ -659,6 +659,105 @@ static std::string GetEndpointFromUrl(std::string const& url) {
 /// `process-utils.js` depends on simple http client error messages.
 ///   this needs to be adjusted if this is ever changed!
 ////////////////////////////////////////////////////////////////////////////////
+namespace {
+
+auto getEndpoint(v8::Isolate* isolate, std::vector<std::string>const& endpoints
+                ,std::string& url, std::string& lastEndpoint) -> std::tuple<std::string, std::string, std::string> {
+
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]: Indentation of the whole function body is wrong.

File: lib/V8/v8-utils.cpp [hunk 5]
```diff
@@ -659,6 +659,105 @@ static std::string GetEndpointFromUrl(std::string const& url) {
...
+        }
+        endpoint = "ssl:" + endpoint;
+      }
+    } else {
+      std::string msg("unsupported URL specified: '");
+      msg.append(url).append("'");
+      return {"","", std::move(msg)};
+    }
+    return {std::move(endpoint), std::move(relative), ""};
+  };
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]:
```suggestion
  }
```

**Comment by Simran-B** [comment]
Copied from original #11232 which was changed so something else since then:

---

Requests may get redirected to a different endpoint. Let the user know the endpoint in case it is disallowed by endpoint black/whitelist for easier diagnosis, e.g. `--javascript.endpoints-whitelist "^ssl://arangodb\.com:443$"` and the following request:

    // redirects to disallowed www subdomain
    require("internal").download("https://arangodb.com")


Related: https://github.com/arangodb/docs/issues/92

```
simran@c9 ~/o/w/ArangoDB> ./build/bin/arangosh --javascript.endpoints-whitelist "ssl://arangodb\.com:443\$"

127.0.0.1:8529@_system> require("internal").download("https://arangodb.com/file.zip")
JavaScript exception: ArangoError 11: not allowed to connect to this endpoint: ssl://www.arangodb.com:443
!require("internal").download("https://arangodb.com/file.zip")
!                    ^
stacktrace: ArangoError: not allowed to connect to this endpoint: ssl://www.arangodb.com:443
    at <shell command>:1:21
```

---

Done in https://github.com/arangodb/arangodb/pull/11236/commits/912fe72e40d05cbe330618e52eee833f769c029f#diff-cf342f102b1a80552625f5193a113b4bR999 and https://github.com/arangodb/arangodb/pull/11236/commits/eb6d15354b4af1bd6c9721e8216e80352c3bb3a3

File: lib/V8/v8-utils.cpp (hunk 3 above)
Comment (ObiWahn) [reply lib/V8/v8-utils.cpp]: I am not sure the code is just moved up and it was in the original code like that. Maybe h2c does the initial connection over http. Maybe @graetzer has an opinion on this.

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9152/

File: lib/V8/v8-utils.cpp [hunk 6]
```diff
@@ -661,6 +661,105 @@ static std::string GetEndpointFromUrl(std::string const& url) {
...
+    endpoint = GetEndpointFromUrl(url).substr(6);
+    relative = url.substr(6 + endpoint.length());
+
+    if (relative.empty() || relative[0] != '/') {
+      relative = "/" + relative;
+    }
+    if (endpoint.find(':') == std::string::npos) {
+      endpoint.append(":80");
+    }
+    endpoint = "tcp://" + endpoint;
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]: shouldn't this be 
```suggestion
    endpoint = "ssl://" + endpoint;
```

Comment (ObiWahn) [reply lib/V8/v8-utils.cpp]: I am not sure the code is just moved up and it was in the original code like that. Maybe h2c does the initial connection over http. Maybe @graetzer has an opinion on this.

Comment (ObiWahn) [reply lib/V8/v8-utils.cpp]: @jsteemann 
```
simon:arango:  15:48
kp kann man wahrscheinlich ssl:// draus machen wenn das h2s vorher war
15:48
h2s ist eh nur unser propiertäres ding um http2 einzuschalten
```
I am not sure what to make out of this? Do you want me switch it to `ssl`?

Comment (jsteemann) [reply lib/V8/v8-utils.cpp]: Yes, please!

Comment (ObiWahn) [reply lib/V8/v8-utils.cpp]: Should the port be changed as well?

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9160/

File: lib/V8/v8-utils.cpp [hunk 7]
```diff
@@ -712,7 +712,7 @@ auto getEndpoint(v8::Isolate* isolate, std::vector<std::string> const& endpoints
     if (endpoint.find(':') == std::string::npos) {
       endpoint.append(":80");
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]: shouldn't we also move to port 443 here?

Comment (ObiWahn) [reply lib/V8/v8-utils.cpp]: https://github.com/arangodb/arangodb/pull/11236#discussion_r397990258

File: lib/V8/v8-utils.cpp (hunk 6 above)
Comment (jsteemann) [reply lib/V8/v8-utils.cpp]: Yes

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9161/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9162/

**Pull request merged**